[language-agnostic-wg] first attempt at a language neutral explanation of the Tracer API (#354)
open-telemetry/opentelemetry-specification: specification/glossary.md

# Glossary

## User Definitions

- <a name="api-developer"></a>API Developer: A developer working on the API
  library for a particular language.
- <a name="sdk-developer"></a>SDK Developer: A developer working on implementing
  the logic either defined in the official SDK specification or a third party
  implementation. The implementation must be used only through the API by a
  library developer or end user.
- <a name="library-developer"></a>Library Developer: This is a developer working
  on code that will be used by others. They are not creating a final
  deployable artifact and must only rely on the OpenTelemetry API as a
  dependency of their library.

> - <a name="end-user"></a>End User: An end user is responsible for code that

Author: Ok, will fix.

  becomes a deployable artifact to run with some configuration by
  an operator. The end user's project may depend on third party libraries that
  have been instrumented and may include its own libraries, making the end
  user potentially a library developer as well. But only the end user, or

>   operator, may include an OpenTelemetry SDK implementation as a dependency and

Author: I guess I can change it to SHOULD?

>   operator, may include an OpenTelemetry SDK implementation as a dependency and

Author: That is what I did. Does that work?

>   operator, may include an OpenTelemetry SDK implementation as a dependency and

Reviewer: Fine with me! If we want to have that imposed as an actual restriction it would have to be placed somewhere other than the glossary.

# Glossary

## User Definitions

- <a name="app-developer"></a>Application Developer: An application developer is
  responsible for code that becomes a deployable artifact to run with some
  configuration by an operator. The application developer's project may depend
  on third party libraries that have been instrumented and may include its own

>   libraries, making the application developer potentially a library developer as

Author: No, I don't think that was the intent here. It is definitely a confusing sentence now and maybe should just be removed.

  well. But only the end user, or operator, should include an OpenTelemetry SDK

>   implementation as a dependency and configure, either through code or

Reviewer 2: This sentence seems out of place for a user glossary definition for an application developer. It seems prescriptive of the operation of OpenTelemetry instrumented software rather than defining a user category.

Does it belong somewhere in the specification or maybe a Roles section?

Author: I could see all of this instead moved to a roles section of the spec and removing the glossary. I dont't think these roles being in a glossary can say much if they don't describe these points, so if a Roles section is to be done instead I'd say it should all be moved.

  configuration files loaded by the program, the `TracerProvider` used by all libraries
  within the final program.
- <a name="library-developer"></a>Library Developer: This is a developer working

>   on code that will be used by others. They are not creating a final deployable

Reviewer 2: We should be specific what we mean by others use here.

```suggestion
  on code that will be integrated into other code. They are not creating a final deployable
```

# Glossary

## User Definitions

> - <a name="api-developer"></a>API Developer: A developer working on the

Reviewer 3: Glossaries are usually done in alphabetical order. If we want custom order, I would start with AppDev (the ultimate end user), then the LibDev (the actual user of the API), followed by the other definitions.

  OpenTelemetry API library for a particular language.

> - <a name="sdk-developer"></a>SDK Developer: A developer working on implementing

Reviewer 3: why use explicit HTML instead of section headers?

Author: It was the only way I knew to be able to have links to bullets. I can make each a section instead, just thought a list was more of a glossary layout.

Reviewer 4: I've seen it done both ways. The anchor tag doesn't particularly bother me but some people disagree.

  library developer or application developer.
- <a name="library-developer"></a>Library Developer: This is a developer working
  on code that will be used by others. They are not creating a final deployable
  artifact and must only rely on the OpenTelemetry API as a dependency of their
  library. The library may have the express purpose of making another library
  observable (such libraries are called "integrations") or they may develop a
  library with any other purpose that has observability built in (e.g., a
  database client library that creates Spans itself when making database calls).
- <a name="app-developer"></a>Application Developer: An application developer is
  responsible for code that becomes a deployable artifact to run with some
  configuration by an operator. The application developer's project may depend
  on third party libraries that have been instrumented and may include its own
  libraries, making the application developer potentially a library developer as
  well. But only the end user, or operator, should include an OpenTelemetry SDK
  implementation as a dependency and configure, either through code or

>   configuration files loaded by the program, the `Tracer` used by all libraries

Reviewer 3: ```suggestion
  configuration files loaded by the program, the `TracerProvider` used by all libraries
```

  within the final program.

> - <a name="operator"></a>Operator: The operator SHOULD be able to select a

Reviewer 3: I think there is a discrepancy in the use of the terminology in the first document, which only mentions AppDevs as those who "choose the SDK".

Author: The operator is never mentioned in the api spec, but it is still a role that must be covered, just one that I haven't adding anything to the spec to cover.

Author: I changed the operator definition to only be `The person deploying and maintaining the
  deployable artifact.` for the time being. Is that good or should I remove Operator entirely until the role is mentioned in the spec somewhere?
